crates/biome_js_analyze/src/semantic_analyzers/correctness/use_hook_at_top_level.rs

```diff
@@ -100,14 +121,159 @@ fn enclosing_function_if_call_is_at_top_level(call: &JsCallExpression) -> Option
         .and_then(|body| body.parent::<AnyJsFunction>())
 }
 
+/// Model for tracking which function calls are preceeded by an early return.
+#[derive(Clone, Default)]
+struct EarlyReturnsModel(HashMap<JsCallExpression, TextRange>);
+
+impl Deref for EarlyReturnsModel {
+    type Target = HashMap<JsCallExpression, TextRange>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for EarlyReturnsModel {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+#[derive(Default)]
+struct EarlyReturnDetectionVisitor {
+    early_returns: EarlyReturnsModel,
+    stack: Vec<EarlyReturnDetectionVisitorStackEntry>,
+}
+
+#[derive(Default)]
+struct EarlyReturnDetectionVisitorStackEntry {
+    early_return: Option<TextRange>,
+}
+
+impl Visitor for EarlyReturnDetectionVisitor {
+    type Language = JsLanguage;
+
+    fn visit(
+        &mut self,
+        event: &WalkEvent<SyntaxNode<Self::Language>>,
+        _ctx: VisitorContext<Self::Language>,
+    ) {
+        match event {
+            WalkEvent::Enter(node) => {
+                if AnyFunctionLike::can_cast(node.kind()) {
+                    self.stack
+                        .push(EarlyReturnDetectionVisitorStackEntry::default());
+                }
+            }
+            WalkEvent::Leave(node) => {
+                if AnyFunctionLike::can_cast(node.kind()) {
+                    self.stack.pop();
+                    return;
+                }
+
+                if let Some(entry) = self.stack.last_mut() {
+                    if let Some(early_return) = JsReturnStatement::cast_ref(node) {
+                        entry.early_return = Some(early_return.syntax().text_range());
```

Comment:
I think we could avoid the cast:
```suggestion
                    if JsReturnStatement::can_cast(node.kind()) {
                        entry.early_return = Some(node.text_range());
```

Reply:
Good catch :+1: